File: backend/infrastructure/agents/agent_factory.py
"""
Agent Factory - Dynamic agent creation based on task analysis.
Based on ADR-002: Zero hardcoded agent classes, all specialization is data-driven.
"""
import logging
from typing import Dict, List
from domain.entities import TaskAnalysis, TaskCapabilities
from infrastructure.tools.tool_registry import ToolRegistry
from infrastructure.agents.plan_execute_agent import PlanExecuteAgent


logger = logging.getLogger(__name__)


class AgentFactory:
    """
    Factory for creating agents dynamically based on task requirements.
    
    Key Principle: No hardcoded agent classes. Agents are created by:
    1. Analyzing task requirements (what capabilities needed)
    2. Getting tools that match capabilities
    3. Generating specialized system prompt
    4. Creating generic PlanExecuteAgent with specific tools/prompt
    """
    
    def __init__(self, tool_registry: ToolRegistry, llm_provider: any):
        self.tool_registry = tool_registry
        self.llm_provider = llm_provider
    
    async def create_agent_for_task(
        self,
        task_description: str,
        context: Dict = None
    ) -> PlanExecuteAgent:
        """
        Create an agent optimized for the specific task.
        
        Process:
        1. Analyze task to determine required capabilities
        2. Select tools that match capabilities
        3. Generate task-specific system prompt
        4. Create PlanExecuteAgent with tools and prompt
        
        Args:
            task_description: Natural language task description
            context: Additional context for agent creation
        
        Returns:
            Configured PlanExecuteAgent ready to execute
        """
        logger.info(f"Creating agent for task: {task_description[:50]}...")
        
        # Step 1: Analyze task requirements
        analysis = await self._analyze_task(task_description, context or {})
        logger.info(f"Task requires capabilities: {analysis.capabilities.primary_capabilities}")
        
        # Step 2: Get matching tools
        tools = self._select_tools_for_capabilities(analysis.capabilities)
        logger.info(f"Selected {len(tools)} tools for task")
        
        # Step 3: Create specialized agent
        agent = PlanExecuteAgent(
            tool_registry=self.tool_registry,
            llm_provider=self.llm_provider,
            max_replan_attempts=2,
            step_timeout_seconds=300
        )
        
        return agent
    
    async def _analyze_task(
        self, task_description: str, context: Dict
    ) -> TaskAnalysis:
        """
        Analyze task to determine requirements.
        
        Uses LLM to understand:
        - What capabilities are needed
        - Complexity level
        - Whether mobile or desktop execution is best
        """
        # Prompt LLM to analyze task
        available_capabilities = self.tool_registry.get_all_capabilities()
        
        prompt = f"""Analyze this task and determine what capabilities are needed:

Task: {task_description}

Available Capabilities:
{', '.join(available_capabilities)}

Provide analysis as JSON:
{{
  "primary_capabilities": ["CAPABILITY1", "CAPABILITY2"],
  "complexity": "simple|moderate|complex",
  "requires_mobile": true|false,
  "requires_desktop": true|false,
  "estimated_duration_minutes": 10
}}
"""
        
        # For now, return default analysis
        # Real implementation would parse LLM response
        return TaskAnalysis(
            task_description=task_description,
            capabilities=TaskCapabilities(
                primary_capabilities=["CODE_GENERATION", "FILE_WRITE"],
                secondary_capabilities=[],
                complexity="moderate",
                estimated_steps=3,
                requires_desktop=True
            ),
            recommended_execution="remote",
            reasoning="Task requires code generation which is desktop capability",
            estimated_duration_minutes=10
        )
    
    def _select_tools_for_capabilities(
        self, capabilities: TaskCapabilities
    ) -> List[str]:
        """Select tools that match required capabilities"""
        all_capabilities = (
            capabilities.primary_capabilities + 
            capabilities.secondary_capabilities
        )
        
        selected_tools = set()
        for capability in all_capabilities:
            tool_names = self.tool_registry.list_tools_by_capability(capability)
            selected_tools.update(tool_names)
        
        return list(selected_tools)
    
    def _generate_system_prompt(
        self, task_description: str, capabilities: TaskCapabilities, tools: List[str]
    ) -> str:
        """Generate specialized system prompt for the agent"""
        return f"""You are an AI agent specialized in: {', '.join(capabilities.primary_capabilities)}.

Your task: {task_description}

Available tools: {', '.join(tools)}

Guidelines:
1. Break down the task into clear, executable steps
2. Use the most appropriate tools for each step
3. Verify each step completed successfully before proceeding
4. If something fails, analyze the error and adjust your plan

Be thorough, efficient, and always explain your reasoning.
"""

File: micro/lib/infrastructure/ai/temp_agent_swarm_backup/agent/agent_factory.dart
import 'package:logger/logger.dart';

import 'models/agent_models.dart';
import 'plan_execute_agent.dart';
import 'tools/tool_registry.dart';

/// Factory for dynamically creating agents
///
/// This factory:
/// 1. Analyzes incoming tasks to understand their complexity and requirements
/// 2. Selects appropriate tools based on task analysis
/// 3. Creates and configures PlanExecuteAgent instances
/// 4. Decides whether task should run locally or remotely
/// 5. Provides execution context (local/remote hybrid)
class AgentFactory {
  final LanguageModel model;
  final ToolRegistry toolRegistry;
  final Logger logger;
  final bool enableRemoteDelegation;

  AgentFactory({
    required this.model,
    required this.toolRegistry,
    required this.logger,
    this.enableRemoteDelegation = true,
  });

  /// Analyze a task to determine its requirements
  Future<TaskAnalysis> analyzeTask(String taskDescription) async {
    logger.d('Analyzing task: $taskDescription');

    // Build analysis prompt
    final availableCapabilities =
        toolRegistry.getAllCapabilities().toList().join(', ');

    final analysisPrompt = '''Analyze this task and determine:
1. Estimated complexity (1-10)
2. Required capabilities from available set: $availableCapabilities
3. Whether it should run on remote server or local device
4. Brief reasoning

Task: $taskDescription

Respond in this JSON format:
{
  "complexity": <number 1-10>,
  "requiredCapabilities": ["capability1", "capability2"],
  "shouldRunRemotely": <true|false>,
  "reasoning": "<explanation>"
}''';

    try {
      final response = await model.invoke(analysisPrompt);
      final analysis = _parseTaskAnalysis(response.content, taskDescription);
      return analysis;
    } catch (e) {
      logger.e('Task analysis failed: $e');
      // Return conservative default analysis
      return TaskAnalysis(
        taskDescription: taskDescription,
        estimatedComplexity: 5,
        requiredCapabilities: [],
        shouldRunRemotely: false,
        reasoning: 'Analysis failed, using defaults',
      );
    }
  }

  /// Create an agent for the given task
  ///
  /// Returns null if the required tools are not available
  PlanExecuteAgent? createAgent(
    String taskDescription, {
    Duration? stepTimeout,
    int? maxReplanAttempts,
  }) {
    logger.d('Creating agent for task: $taskDescription');

    // Analyze task
    // NOTE: In production, this would be async, but we'd need to redesign this method
    // For now, we create agent optimistically and let plan creation handle capability validation

    return PlanExecuteAgent(
      model: model,
      toolRegistry: toolRegistry,
      logger: logger,
      stepTimeout: stepTimeout ?? const Duration(minutes: 5),
      maxReplanAttempts: maxReplanAttempts ?? 3,
    );
  }

  /// Create an agent async with full analysis
  ///
  /// This is the recommended method that performs full task analysis before
  /// creating the agent
  Future<PlanExecuteAgent?> createAgentAsync(
    String taskDescription, {
    Duration? stepTimeout,
    int? maxReplanAttempts,
  }) async {
    logger.d('Creating agent (async) for task: $taskDescription');

    // Analyze task
    final analysis = await analyzeTask(taskDescription);
    logger.d(
      'Task analysis: complexity=${analysis.estimatedComplexity}, '
      'remote=${analysis.shouldRunRemotely}',
    );

    // Check if required capabilities are available
    if (!toolRegistry.hasAllCapabilities(analysis.requiredCapabilities)) {
      logger.w(
        'Required capabilities not available. Required: '
        '${analysis.requiredCapabilities}, '
        'Available: ${toolRegistry.getAllCapabilities()}',
      );
      return null;
    }

    // Create agent
    final agent = PlanExecuteAgent(
      model: model,
      toolRegistry: toolRegistry,
      logger: logger,
      stepTimeout: stepTimeout ?? const Duration(minutes: 5),
      maxReplanAttempts: maxReplanAttempts ?? 3,
    );

    logger.d('Agent created successfully');
    return agent;
  }

  /// Create multiple specialized agents for complex tasks
  ///
  /// For tasks that require multiple domains of expertise, this creates
  /// a team of specialized agents that can work together
  Future<List<PlanExecuteAgent>> createAgentTeam(
    String mainTaskDescription,
    List<String> subtasks,
  ) async {
    logger.d('Creating agent team for ${subtasks.length} subtasks');

    final agents = <PlanExecuteAgent>[];

    for (final subtask in subtasks) {
      final agent = await createAgentAsync(subtask);
      if (agent != null) {
        agents.add(agent);
      } else {
        logger.w('Could not create agent for subtask: $subtask');
      }
    }

    logger.d('Agent team created with ${agents.length} agents');
    return agents;
  }

  /// Estimate task complexity without creating an agent
  ///
  /// Useful for deciding whether to execute locally or delegate to remote
  Future<int> estimateComplexity(String taskDescription) async {
    final analysis = await analyzeTask(taskDescription);
    return analysis.estimatedComplexity;
  }

  /// Determine if task should run remotely
  ///
  /// Used by mobile app to decide between local execution and remote delegation
  Future<bool> shouldExecuteRemotely(String taskDescription) async {
    if (!enableRemoteDelegation) return false;

    try {
      final analysis = await analyzeTask(taskDescription);
      return analysis.shouldRunRemotely;
    } catch (e) {
      logger.e('Could not determine execution context: $e');
      return false;
    }
  }

  /// Get planning context for an agent
  ///
  /// This provides all information needed for the agent's planning phase
  PlanningContext getPlanningContext(String taskDescription) {
    return PlanningContext(
      taskDescription: taskDescription,
      availableTools: toolRegistry.getAllMetadata(),
      availablePermissions: _getAvailablePermissions(),
      environmentInfo: _getEnvironmentInfo(),
    );
  }

  /// Parse task analysis from LLM response
  TaskAnalysis _parseTaskAnalysis(
    String response,
    String originalTask,
  ) {
    try {
      // Extract JSON from response
      final jsonMatch = RegExp(r'\{.*\}', dotAll: true).firstMatch(response);
      if (jsonMatch == null) {
        throw Exception('No JSON found in response');
      }

      // Parse JSON - would use json.decode in production with proper error handling
      // For now, using simple pattern matching as fallback

      final jsonStr = jsonMatch.group(0)!;

      // Extract complexity
      int complexity = 5;
      final complexityMatch =
          RegExp(r'"complexity"\s*:\s*(\d+)').firstMatch(jsonStr);
      if (complexityMatch != null) {
        complexity = int.parse(complexityMatch.group(1)!);
      }

      // Extract remote flag
      bool remote = false;
      if (jsonStr.contains('"shouldRunRemotely": true')) {
        remote = true;
      }

      // Extract required capabilities
      final List<String> capabilities = [];
      final capMatch =
          RegExp(r'"requiredCapabilities"\s*:\s*\[(.*?)\]').firstMatch(jsonStr);
      if (capMatch != null) {
        final capStr = capMatch.group(1)!;
        final caps = capStr.split(',');
        for (final cap in caps) {
          final cleaned = cap.trim().replaceAll('"', '');
          if (cleaned.isNotEmpty) {
            capabilities.add(cleaned);
          }
        }
      }

      return TaskAnalysis(
        taskDescription: originalTask,
        estimatedComplexity: complexity,
        requiredCapabilities: capabilities,
        shouldRunRemotely: remote && enableRemoteDelegation,
        reasoning: 'Task analysis complete',
      );
    } catch (e) {
      logger.w('Failed to parse task analysis: $e');
      return TaskAnalysis(
        taskDescription: originalTask,
        estimatedComplexity: 5,
        requiredCapabilities: [],
        shouldRunRemotely: false,
        reasoning: 'Parse failed: $e',
      );
    }
  }

  /// Get available permissions for context
  List<String> _getAvailablePermissions() {
    // In production, this would check actual device permissions
    return [
      'camera',
      'microphone',
      'location',
      'contacts',
      'calendar',
      'files',
      'sensors',
    ];
  }

  /// Get environment information for planning context
  Map<String, dynamic> _getEnvironmentInfo() {
    return {
      'platform': 'flutter',
      'toolCount': toolRegistry.toolCount,
      'availableCapabilities': toolRegistry.getAllCapabilities().toList(),
      'timestamp': DateTime.now().toIso8601String(),
    };
  }
}
